src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,732 @@
+        // TODO: come up with some better heuristic/budget
+        JITDUMP("impExpandHalfConstEquals: method has too many BBs (>20) - not profitable to expand.\n");
+        return nullptr;
+    }
+
+    GenTree* elementsCount = gtNewIconNode(len);
+    GenTree* lenCheckNode;
+    if (len == 0)
+    {
+        if (startsWith)
+        {
+            // Any string starts with ""
+            return gtNewTrue();
+        }
+
+        // For zero length we don't need to compare content, the following expression is enough:
+        //
+        //   varData != null && lengthFld == 0
+        //
+        lenCheckNode = gtNewOperNode(GT_EQ, TYP_INT, lengthFld, elementsCount);
+    }
+    else
+    {
+        assert(cnsData != nullptr);
+
+        GenTree* indirCmp = nullptr;
+        if (len < 8) // SWAR impl supports len == 8 but we'd better give it to SIMD
+        {
+            indirCmp = impExpandHalfConstEqualsSWAR(gtClone(data), cnsData, len, dataOffset);
+        }
+        else if (len <= 32)
+        {
+            indirCmp = impExpandHalfConstEqualsSIMD(gtClone(data), cnsData, len, dataOffset);
+        }
+
+        if (indirCmp == nullptr)
+        {
+            JITDUMP("unable to compose indirCmp\n");
+            return nullptr;
+        }
+
+        GenTreeColon* lenCheckColon = gtNewColonNode(TYP_INT, indirCmp, gtNewFalse());
+
+        // For StartsWith we use GT_GE, e.g.: `x.Length >= 10`
+        lenCheckNode =
+            gtNewQmarkNode(TYP_INT, gtNewOperNode(startsWith ? GT_GE : GT_EQ, TYP_INT, lengthFld, elementsCount),
+                           lenCheckColon);
+    }
+
+    GenTree* rootQmark;
+    if (checkForNull)
+    {
+        // varData == nullptr
+        GenTreeColon* nullCheckColon = gtNewColonNode(TYP_INT, lenCheckNode, gtNewFalse());
+        rootQmark = gtNewQmarkNode(TYP_INT, gtNewOperNode(GT_NE, TYP_INT, data, gtNewNull()), nullCheckColon);
+    }
+    else
+    {
+        // no nullcheck, just "obj.Length == len && (SWAR or SIMD)"
+        rootQmark = lenCheckNode;
+    }
+
+    return rootQmark;
+}
+
+//------------------------------------------------------------------------
+// impGetStrConFromSpan: Try to obtain string literal out of a span:
+//  var span = "str".AsSpan();
+//  var span = (ReadOnlySpan<char>)"str"
+//
+// Arguments:
+//    span - ReadOnlySpan tree
+//
+// Returns:
+//    GenTreeStrCon node or nullptr
+//
+GenTreeStrCon* Compiler::impGetStrConFromSpan(GenTree* span)
+{
+    GenTreeCall* argCall = nullptr;
+    if (span->OperIs(GT_RET_EXPR))
+    {
+        argCall = span->AsRetExpr()->gtInlineCandidate->AsCall();
+    }
+    else if (span->OperIs(GT_CALL))
+    {
+        argCall = span->AsCall();
+    }
+
+    if (argCall != nullptr)
+    {
+        NamedIntrinsic ni = lookupNamedIntrinsic(argCall->gtCallMethHnd);
+        if ((ni == NI_System_MemoryExtensions_AsSpan) || (ni == NI_System_String_op_Implicit))
+        {
+            if (argCall->gtCallArgs->GetNode()->OperIs(GT_CNS_STR))
```

Comment:
We may want to assert here that there is just one method argument.

There are multiple overloads of MemoryExtension.AsSpan. This only works for the one that does not take any indices. It is the only marked as intrinsic today, so we should be fine. Just in case somebody tries to mark the other one as intrinsic too...

Reply:
Good point! yeah, a randomly inserted [Intrinsic] in *.cs can indeed break this code